tools/crm_attribute.c

```diff
@@ -30,357 +30,310 @@
-            "to reboot (normally invoked from an OCF resource agent)",
-        pcmk__option_default
-    },
-    {
-        "set-name", required_argument, NULL, 's',
-        "(Advanced) Operate on instance of specified attribute that is "
-            "within set with this XML ID",
-        pcmk__option_default
-    },
-    {
-        "id", required_argument, NULL, 'i',
-        "\t(Advanced) Operate on instance of specified attribute with this "
-            "XML ID",
-        pcmk__option_default
-    },
+static gboolean
+delete_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **error) {
+    options.command = 'D';
 
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "\nCommands:", pcmk__option_default
-    },
-    {
-        "query", no_argument, NULL, 'G',
-        "\tQuery the current value of the attribute/option",
-        pcmk__option_default
-    },
-    {
-        "update", required_argument, NULL, 'v',
-        "Update the value of the attribute/option", pcmk__option_default
-    },
-    {
-        "delete", no_argument, NULL, 'D',
-        "\tDelete the attribute/option", pcmk__option_default
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "\nAdditional Options:", pcmk__option_default
-    },
-    {
-        "node", required_argument, NULL, 'N',
-        "Set a node attribute for named node (instead of a cluster option). "
-            "See also: -l",
-        pcmk__option_default
-    },
-    {
-        "type", required_argument, NULL, 't',
-        "Which part of the configuration to update/delete/query the option in",
-        pcmk__option_default
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "\t\t\tValid values: crm_config, rsc_defaults, op_defaults, tickets",
-        pcmk__option_default
-    },
-    {
-        "lifetime", required_argument, NULL, 'l',
-        "Lifetime of the node attribute", pcmk__option_default
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "\t\t\tValid values: reboot, forever", pcmk__option_default
-    },
-    {
-        "utilization", no_argument, NULL, 'z',
-        "Set an utilization attribute for the node.", pcmk__option_default
-    },
-    {
-        "default", required_argument, NULL, 'd',
-        "(Advanced) Default value to display if none is found in configuration",
-        pcmk__option_default
-    },
-    {
-        "inhibit-policy-engine", no_argument, NULL, '!',
-        NULL, pcmk__option_hidden
-    },
+    if (options.attr_value) {
+        free(options.attr_value);
+    }
+
+    options.attr_value = NULL;
+    return TRUE;
+}
+
+static gboolean
+promotion_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **error) {
+    options.promotion_score = TRUE;
+
+    if (options.attr_name) {
+        g_free(options.attr_name);
+    }
+
+    options.attr_name = g_strdup(pcmk_promotion_score_name(optarg));
```

Comment:
pcmk_promotion_score_name() can be NULL (if no optional argument was given and OCF_RESOURCE_INSTANCE is not set in environment, which would be a usage error), otherwise it returns dynamic memory that will need to be free()'d after g_strdup()'ing it. Note that there's a check for the NULL case later, so either that could be moved here or this could set NULL.

Reply:
Alright, I think I've got this one straightened out.